src/hammer-vlsi/par/openroad/__init__.py

```diff
@@ -84,148 +181,959 @@ def tool_config_prefix(self) -> str:
+        """
+        pre_filters = optional_map(corner, lambda c: [self.filter_for_mmmc(voltage=c.voltage,temp=c.temp)])  # type: Optional[List[Callable[[hammer_tech.Library],bool]]]
+        lib_args = self.technology.read_libs([hammer_tech.filters.timing_lib_with_ecsm_filter], hammer_tech.HammerTechnologyUtils.to_plain_item, extra_pre_filters=pre_filters)
+        return " ".join(lib_args)
+
+    def read_lef(self) -> bool:
+        # OpenROAD names the LEF libraries by filename:
+        #   foo.tlef and foo.lef evaluate to the same library "foo"
+        #   solution: copy foo.lef to foo1.lef
+        lef_files = self.technology.read_libs([
+            hammer_tech.filters.lef_filter
+        ], hammer_tech.HammerTechnologyUtils.to_plain_item)
+        if self.hierarchical_mode.is_nonleaf_hierarchical():
+            ilm_lefs = list(map(lambda ilm: ilm.lef, self.get_input_ilms()))
+            lef_files.extend(ilm_lefs)
+        lef_file_libnames=[]
+        unique_id=0
+        for lef_file in lef_files:
+            lef_file_name=lef_file.split('/')[-1]
+            lef_file_libname=''.join(lef_file_name.split('.')[:-1])
+            lef_file_ext=lef_file_name.split('.')[-1]
+            if lef_file_libname in lef_file_libnames:
+                lef_file_libname=f"{lef_file_libname}_{unique_id}.{lef_file_ext}"
+                new_lef_file=f"{self.run_dir}/{lef_file_libname}"
+                with open(lef_file,'r') as f_old:
+                    with open(new_lef_file,'w') as f_new:
+                        for line in f_old: 
+                            f_new.write(line)
+                lef_file=new_lef_file
+                unique_id+=1
+            lef_file_libnames.append(lef_file_libname)
+            self.append(f"read_lef {lef_file}")
+        self.append("")
+        return True
+
+    def read_liberty(self) -> bool:
+        corners = self.get_mmmc_corners()  # type: List[MMMCCorner]
+        if corners:
+            corner_names = []
+            for corner in corners:
+                # Setting up views for all defined corner types: setup, hold, extra
+                if corner.type is MMMCCornerType.Setup:
+                    corner_name="setup"
+                elif corner.type is MMMCCornerType.Hold:
+                    corner_name="hold"
+                elif corner.type is MMMCCornerType.Extra:
+                    corner_name="extra"
+                else:
+                    raise ValueError("Unsupported MMMCCornerType")
+                corner_names.append(corner_name)
+            
+            self.append(f"define_corners {' '.join(corner_names)}")
+            for corner,corner_name in zip(corners,corner_names):
+                lib_files=self.get_timing_libs(corner)
+                for lib_file in lib_files.split():
+                    self.verbose_append(f"read_liberty -corner {corner_name} {lib_file}")
+        self.append("")
+
+    def convert_units(self,prefix) -> str:
```

Comment:
I renamed this function to clarify...but openroad errors when reading something like "1000ps" because it expects "1ns" (the error msg says it doesn't recognize the unit "00ps"). So the function scale_units_1000x_down (renamed from convert_units) scales this down to do the conversion. Maybe I could add this to units.py eventually, but it doesn't seem useful in any other scenarios.